File: frontend/src/services/geopayService.ts
import api from './api';

export interface RewardsBalance {
  balance: number;
  discountBalance: number;
  lifetimeEarned: number;
  lifetimeRedeemed: number;
  redeemableDiscount: number;
}

export interface RewardTransaction {
  id: string;
  userId: string;
  points: number;
  type: 'earned' | 'redeemed' | 'expired';
  cashValue: number;
  description?: string;
  referenceId?: string;
  createdAt: string;
}

export interface ReferralInfo {
  referralCode: string;
  totalReferrals: number;
  pendingReferrals: number;
}

export interface ReferralEntry {
  id: string;
  referralCode: string;
  referredEmail?: string;
  status: 'pending' | 'rewarded' | 'expired';
  rewardAmount: number;
  createdAt: string;
}

export async function getRewardsBalance(): Promise<RewardsBalance> {
  const response = await api.get<RewardsBalance>('/geopay/rewards/balance');
  return response.data;
}

export async function getRewardHistory(): Promise<RewardTransaction[]> {
  const response = await api.get<RewardTransaction[]>('/geopay/rewards/history');
  return response.data;
}

export async function redeemPoints(points: number): Promise<{
  pointsRedeemed: number;
  discountPesos: number;
  newBalance: number;
  discountBalance: number;
}> {
  const response = await api.post('/geopay/rewards/redeem', { points });
  return response.data;
}

export async function consumeDiscount(discountPesos: number): Promise<{
  consumed: number;
  remainingDiscount: number;
}> {
  const response = await api.post('/geopay/rewards/consume-discount', { discountPesos });
  return response.data;
}

export async function getReferralCode(): Promise<ReferralInfo> {
  const response = await api.get<ReferralInfo>('/geopay/referral/code');
  return response.data;
}

export async function registerReferral(code: string, email?: string): Promise<{ success: boolean }> {
  const response = await api.post('/geopay/referral/register', { code, email });
  return response.data;
}

export async function getReferralHistory(): Promise<ReferralEntry[]> {
  const response = await api.get<ReferralEntry[]>('/geopay/referral/history');
  return response.data;
}

File: flutter_geobite/lib/services/geopay_service.dart
import '../core/api_client.dart';

class GeopayService {
  /// Get current customer's rewards balance (points, discount balance, lifetime earned/redeemed)
  Future<Map<String, dynamic>> getRewardsBalance() async {
    try {
      final response = await apiClient.dio.get('/geopay/rewards/balance');
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load rewards balance: $e');
    }
  }

  /// Get rewards points history
  Future<List<Map<String, dynamic>>> getRewardHistory() async {
    try {
      final response = await apiClient.dio.get('/geopay/rewards/history');
      final data = response.data;
      final List list = data is List ? data : [];
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load reward history: $e');
    }
  }

  /// Redeem loyalty points for discount balance (10 points = ₱1 discount)
  Future<Map<String, dynamic>> redeemPoints(int points) async {
    try {
      final response = await apiClient.dio.post('/geopay/rewards/redeem', data: {
        'points': points,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to redeem points: $e');
    }
  }

  /// Consume discount pesos
  Future<Map<String, dynamic>> consumeDiscount(double discountPesos) async {
    try {
      final response = await apiClient.dio.post('/geopay/rewards/consume-discount', data: {
        'discountPesos': discountPesos,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to consume discount: $e');
    }
  }

  /// Get referral code, total, and pending referrals
  Future<Map<String, dynamic>> getReferralCode() async {
    try {
      final response = await apiClient.dio.get('/geopay/referral/code');
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load referral code: $e');
    }
  }

  /// Register a referral code
  Future<Map<String, dynamic>> registerReferral(String code, String? email) async {
    try {
      final response = await apiClient.dio.post('/geopay/referral/register', data: {
        'code': code,
        if (email != null) 'email': email,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to register referral code: $e');
    }
  }

  /// Get referral history (referred signups and status)
  Future<List<Map<String, dynamic>>> getReferralHistory() async {
    try {
      final response = await apiClient.dio.get('/geopay/referral/history');
      final data = response.data;
      final List list = data is List ? data : [];
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load referral history: $e');
    }
  }
}

final geopayService = GeopayService();
